src/openvic-simulation/dataloader/NodeTools.cpp

```diff
@@ -286,16 +333,19 @@ node_callback_t NodeTools::expect_length(callback_t<size_t> callback) {
 }
 
 node_callback_t NodeTools::expect_key(std::string_view key, node_callback_t callback, bool* key_found, bool allow_duplicates) {
-	return _expect_type<ast::AbstractListNode>(
-		[key, callback, key_found, allow_duplicates](ast::AbstractListNode const& list_node) -> bool {
+	return _abstract_statement_node_callback(
+		[key, callback, key_found, allow_duplicates](_NodeStatementRange list) -> bool {
 			bool ret = true;
 			size_t keys_found = 0;
-			std::vector<ast::NodeUPtr> const& list = list_node._statements;
-			for (ast::NodeUPtr const& sub_node : list_node._statements) {
-				ast::AssignNode const* assign_node = sub_node->cast_to<ast::AssignNode>();
-				if (assign_node != nullptr && assign_node->_name == key) {
+			for (auto sub_node : list) {
+				auto const* assign_node = dryad::node_try_cast<ast::AssignStatement>(sub_node);
+				std::string_view left;
+				if (!expect_identifier(assign_variable_callback_string(left))(assign_node->left())) {
+					continue;
+				}
+				if (assign_node != nullptr && left == key) {
```

Comment:
`assign_node->left()` is called before the `assign_node != nullptr` check, which could cause a segfault if `sub_node` is not successfully cast to an `ast::AssignStatement`.